File: tools/src/builder.ts
// SPDX-License-Identifier: MPL-2.0

import * as path from "@std/path";
import { PROJECT_ROOT, PATHS } from "./defines.ts";
import {
  createFeatureSymlinks,
  exists,
  Logger,
  runCommandChecked,
  safeRemove,
} from "./utils.ts";
import { writeBuildid2 } from "./update.ts";

const logger = new Logger("builder");

export function packageVersion(): string {
  const pkgPath = path.join(PROJECT_ROOT, "package.json");
  const content = Deno.readTextFileSync(pkgPath);
  return JSON.parse(content).version;
}

type CommandTuple = readonly [readonly string[], string];

export async function runInParallel(commands: CommandTuple[]): Promise<void> {
  const results = commands.map(([cmd, dir]) => {
    logger.info(`Running \`${cmd.join(" ")}\` in \`${dir}\``);
    // Use runCommandChecked (sync) to capture stdout/stderr concisely in Deno
    const res = runCommandChecked(cmd[0], cmd.slice(1), dir);
    return {
      success: res.code === 0,
      cmd,
      dir,
      out: res.stdout,
      err: res.stderr,
    };
  });

  for (const res of results) {
    if (!res.success) {
      throw new Error(
        `Build command \`${res.cmd.join(" ")}\` in \`${res.dir}\` failed\nSTDOUT:\n${res.out}\nSTDERR:\n${res.err}`,
      );
    }
  }
}

export async function run(mode = "dev", buildid2: string): Promise<void> {
  logger.info(`Building features with mode=${mode}`);

  // Ensure buildid2 is written to the expected path so other tools can read it.
  try {
    writeBuildid2(PATHS.buildid2, buildid2);
  } catch (e: any) {
    logger.error(`Failed to write buildid2: ${e?.message ?? e}`);
  }

  const version = packageVersion();

  const devCommands: CommandTuple[] = [
    [
      ["deno", "task", "build", `--env.MODE=${mode}`],
      path.join(PROJECT_ROOT, "bridge/startup"),
    ],
    [
      [
        "deno",
        "task",
        "build",
        `--env.MODE=${mode}`,
        `--env.__BUILDID2__=${buildid2}`,
        `--env.__VERSION2__=${version}`,
      ],
      path.join(PROJECT_ROOT, "bridge/loader-modules"),
    ],
    [
      [
        "deno",
        "run",
        "-A",
        "vite",
        "build",
        "--base",
        "chrome://noraneko/content",
      ],
      path.join(PROJECT_ROOT, "browser-features/chrome"),
    ],
    [
      ["deno", "task", "build", `--env.MODE=${mode}`],
      path.join(PROJECT_ROOT, "bridge/loader-features"),
    ],
    [
      ["deno", "task", "build", `--env.MODE=${mode}`],
      path.join(PROJECT_ROOT, "browser-features/webext-actors"),
    ],
    [
      ["deno", "task", "build"],
      path.join(PROJECT_ROOT, "browser-features/pages-aboutDialog"),
    ],
    [
      ["deno", "task", "build"],
      path.join(PROJECT_ROOT, "browser-features/pages-newtab"),
    ],
    [
      ["deno", "task", "build"],
      path.join(PROJECT_ROOT, "browser-features/settings"),
    ],
  ];

  const prodCommands: CommandTuple[] = [
    [
      ["deno", "task", "build", `--env.MODE=${mode}`],
      path.join(PROJECT_ROOT, "bridge/startup"),
    ],
    [
      [
        "deno",
        "run",
        "-A",
        "vite",
        "build",
        "--base",
        "chrome://noraneko/content",
      ],
      path.join(PROJECT_ROOT, "browser-features/chrome"),
    ],
    [
      [
        "deno",
        "task",
        "build",
        `--env.MODE=${mode}`,
        `--env.__BUILDID2__=${buildid2}`,
        `--env.__VERSION2__=${version}`,
      ],
      path.join(PROJECT_ROOT, "bridge/loader-modules"),
    ],
    [
      ["deno", "task", "build", `--env.MODE=${mode}`],
      path.join(PROJECT_ROOT, "bridge/loader-features"),
    ],
    [
      ["deno", "task", "build", `--env.MODE=${mode}`],
      path.join(PROJECT_ROOT, "browser-features/webext-actors"),
    ],
    [
      ["deno", "task", "build"],
      path.join(PROJECT_ROOT, "browser-features/pages-aboutDialog"),
    ],
    [
      ["deno", "task", "build"],
      path.join(PROJECT_ROOT, "browser-features/pages-newtab"),
    ],
    [
      ["deno", "task", "build"],
      path.join(PROJECT_ROOT, "browser-features/settings"),
    ],
  ];

  if (mode.startsWith("dev")) {
    await runInParallel(devCommands);
  } else {
    await runInParallel(prodCommands);
  }

  if (mode.startsWith("production")) {
    const dirPath = "_dist/noraneko";
    try {
      if (exists(dirPath)) {
        safeRemove(dirPath);
      }
    } catch {}
    Deno.mkdirSync(dirPath);

    createFeatureSymlinks(dirPath);
  }

  logger.success("Build complete.");
}

File: tools/lib/builder.rb
# SPDX-License-Identifier: MPL-2.0

require "fileutils"
require "json"
require_relative "defines"
require_relative "update"
require_relative "utils"

module FelesBuild
  # 出来上がりを作る段。vite を library として使うのではなく、各 package の
  # `deno task build` を順に起こすだけ(束ねるのは Deno のほうが得意なので、そこは任せる)。
  module Builder
    LOGGER = Utils::Logger.new("builder")

    def self.package_version
      JSON.parse(File.read(Defines.in_root("package.json")))["version"]
    end

    # [走らせるもの, どこで]。dev と production の違いは chrome を先に建てるかどうかだけ。
    def self.commands(mode, buildid2)
      version = package_version
      chrome = [%w[deno run -A vite build --base chrome://noraneko/content], "browser-features/chrome"]
      loader_modules = [
        ["deno", "task", "build", "--env.MODE=#{mode}",
         "--env.__BUILDID2__=#{buildid2}", "--env.__VERSION2__=#{version}"],
        "bridge/loader-modules",
      ]
      startup = [["deno", "task", "build", "--env.MODE=#{mode}"], "bridge/startup"]
      webext_actors = [["deno", "task", "build", "--env.MODE=#{mode}"], "browser-features/webext-actors"]
      plain = %w[
        browser-features/pages-aboutDialog
        browser-features/pages-newtab
        browser-features/settings
      ].map { |dir| [%w[deno task build], dir] }

      middle = mode.start_with?("dev") ? [loader_modules, chrome] : [chrome, loader_modules]
      [startup, *middle, webext_actors, *plain]
    end

    def self.run(mode, buildid2)
      LOGGER.info "Building features with mode=#{mode}"

      begin
        Update.write_buildid2(buildid2)
      rescue => e
        LOGGER.error "Failed to write buildid2: #{e.message}"
      end

      commands(mode, buildid2).each do |command, dir|
        cwd = Defines.in_root(dir)
        LOGGER.info "Running `#{command.join(' ')}` in `#{cwd}`"
        result = Utils.run_checked(*command, cwd: cwd)
        next if result[:success]

        raise "Build command `#{command.join(' ')}` in `#{cwd}` failed\n" \
              "STDOUT:\n#{result[:stdout]}\nSTDERR:\n#{result[:stderr]}"
      end

      if mode.start_with?("production")
        out = Defines.in_root("_dist", "noraneko")
        FileUtils.rm_rf(out)
        FileUtils.mkdir_p(out)
        Utils.create_feature_symlinks(out)
      end

      LOGGER.success "Build complete."
    end
  end
end
